lib/smack/Regions.cpp

```diff
@@ -18,111 +13,27 @@ namespace smack {
 
 const DataLayout *Region::DL = nullptr;
 DSAWrapper *Region::DSA = nullptr;
-// DSNodeEquivs* Region::NEQS = nullptr;
-
-namespace {
-const Function *getFunction(const Value *V) {
-  if (const Instruction *I = dyn_cast<Instruction>(V))
-    return I->getParent()->getParent();
-  else if (const Argument *A = dyn_cast<Argument>(V))
-    return A->getParent();
-  else if (const BasicBlock *BB = dyn_cast<BasicBlock>(V))
-    return BB->getParent();
-
-  // XXX I know this looks bad, but it works for now
-  for (auto U : V->users())
-    return getFunction(U);
-
-  llvm_unreachable("Unexpected value.");
-}
-
-bool isFieldDisjoint(DSAWrapper *DSA, const Value *V, unsigned offset) {
-  if (const GlobalValue *G = dyn_cast<GlobalValue>(V))
-    return DSA->isFieldDisjoint(G, offset);
-  else
-    return DSA->isFieldDisjoint(V, getFunction(V));
-}
-} // namespace
 
 void Region::init(Module &M, Pass &P) {
   DL = &M.getDataLayout();
   DSA = &P.getAnalysis<DSAWrapper>();
 }
 
-namespace {
-unsigned numGlobals(const DSNode *N) {
-  unsigned count = 0;
-
-  // shamelessly ripped from getCaption(..) in lib/DSA/Printer.cpp
-  EquivalenceClasses<const GlobalValue *> *GlobalECs = 0;
-  const DSGraph *G = N->getParentGraph();
-  if (G)
-    GlobalECs = &G->getGlobalECs();
-
-  for (auto i = N->globals_begin(), e = N->globals_end(); i != e; ++i) {
-    count += 1;
-
-    if (GlobalECs) {
-      // Figure out how many globals are equivalent to this one.
-      auto I = GlobalECs->findValue(*i);
-      if (I != GlobalECs->end()) {
-        count +=
-            std::distance(GlobalECs->member_begin(I), GlobalECs->member_end()) -
-            1;
-      }
-    }
-  }
+bool Region::isSingleton(const Value *v, unsigned length) {
+  // TODO can we do something for non-global nodes?
+  auto node = DSA->getNode(v);
```

Comment:
Add `assert(node)` after this line, just as a sanity check.

Reply:
I don't think we need a sanity check here since getNode already gets one upon returning.

Reply:
Wait. I am not sure about this since `getNode` can return `nullptr`.

Reply:
`isSingleton` only gets called once the `representative` is not nullptr, which means that we don't need to check nullptr deference here.